### PR TITLE
Implement "Forgot Password" #30

### PR DESCRIPTION
**Pull request by fedyafedko**

### CODE REVIEW

File: StudyHub/StudyHub.BLL/Profiles/UserProfile.cs [hunk 1]
```diff
@@ -7,11 +8,16 @@ namespace StudyHub.BLL.Profiles;
...
 public class UserProfile : Profile
 {
-    public UserProfile() 
+    public UserProfile()
     {
         CreateMap<InvitedUserDTO, InvitedUser>();
         CreateMap<RegisterUserDTO, User>()
-            .ForMember(dest => dest.UserName, opt => opt.MapFrom(src => src.Email))
-            .ForMember(dest => dest.EmailConfirmed, opt => opt.MapFrom(_ => true));
+            .ForMember(dest => dest.Email, opt => opt.MapFrom(src => src.Email))
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Profiles/UserProfile.cs:15]: nessesary to map email to email?

File: StudyHub/StudyHub.BLL/Services/Auth/GoogleAuthService.cs [hunk 2]
```diff
@@ -1,45 +1,71 @@
...
     }
-    public async Task<AuthSuccessDTO> GoogleLogin(string authorizationCode)
+
+    public async Task<AuthSuccessDTO> GoogleRegisterAsync(string authorizationCode)
+    {
+        var payload = await GetGooglePayloadAsync(authorizationCode);
+
+        var user = await _userManager.FindByEmailAsync(payload.Email);
+
+        if (user is not null)
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/GoogleAuthService.cs:40]: code style. I mean, meaby you make this part same like in other files to save identity of code style?

File: StudyHub/StudyHub.BLL/Services/Auth/GoogleAuthService.cs [hunk 3]
```diff
@@ -1,45 +1,71 @@
...
+            throw new NotFoundException("User with this email already exists");
+
+        user = _mapper.Map<User>(payload);
+
+        var createdUserResult = await _userManager.CreateAsync(user);
+
+        if (!createdUserResult.Succeeded)
+            throw new UserManagerException("Unable to authenticate given user", createdUserResult.Errors);
+
+        return new AuthSuccessDTO(
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/GoogleAuthService.cs:50]: look at my auth service there are GetAuthTokens metod, I think it better to make like this way

File: StudyHub/StudyHub.BLL/Services/Auth/GoogleAuthService.cs [hunk 4]
```diff
@@ -1,45 +1,71 @@
...
+            await _tokenService.GenerateJwtTokenAsync(user),
+            _tokenService.GenerateRefreshTokenAsync(user));
+    }
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/GoogleAuthService.cs:53]: empty line

File: StudyHub/StudyHub.BLL/Services/Auth/GoogleAuthService.cs [hunk 5]
```diff
@@ -1,45 +1,71 @@
...
-        var paylod = await GetUserInfoAsync(authorizationCode);
+        var paylod = await GetGooglePayloadAsync(authorizationCode);
 
         var user = await _userManager.FindByEmailAsync(paylod.Email);
 
         if (user == null)
             throw new NotFoundException($"Unable to find user by specified email. Email: {user!.Email}");
 
-        return new AuthSuccessDTO(_tokenService.GenerateJwtToken(user, (await _userManager.GetRolesAsync(user)).ToArray()),
+        return new AuthSuccessDTO(
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/GoogleAuthService.cs:63]: same here

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
+﻿using Microsoft.AspNetCore.Identity;
+using StudyHub.BLL.Services.Interfaces.Auth;
+using StudyHub.Common.DTO.AuthDTO;
+using StudyHub.Common.Exceptions;
+using StudyHub.Entities;
+
+namespace StudyHub.BLL.Services.Auth;
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:7]: empty line between namespace

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    private readonly UserManager<User> _userManager;
+    public PasswordService(UserManager<User> userManager)
+    {
+        _userManager = userManager;
+    }
+    public async Task<string> ForgotPassword(ForgotPasswordDTO dto)
+    {
+        var user = await _userManager.FindByEmailAsync(dto.Email);
+
+        if (user == null)
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:19]: I think its better to make like ,,?? throw new ..... ,,

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+using StudyHub.Entities;
+
+namespace StudyHub.BLL.Services.Auth;
+public class PasswordService : IPasswordService
+{
+    private readonly UserManager<User> _userManager;
+    public PasswordService(UserManager<User> userManager)
+    {
+        _userManager = userManager;
+    }
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:14]: empty line

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
+﻿using Microsoft.AspNetCore.Identity;
+using StudyHub.BLL.Services.Interfaces.Auth;
+using StudyHub.Common.DTO.AuthDTO;
+using StudyHub.Common.Exceptions;
+using StudyHub.Entities;
+
+namespace StudyHub.BLL.Services.Auth;
+public class PasswordService : IPasswordService
+{
+    private readonly UserManager<User> _userManager;
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:10]: empty line

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    public PasswordService(UserManager<User> userManager)
+    {
+        _userManager = userManager;
+    }
+    public async Task<string> ForgotPassword(ForgotPasswordDTO dto)
+    {
+        var user = await _userManager.FindByEmailAsync(dto.Email);
+
+        if (user == null)
+            throw new NotFoundException($"User with this email does not exist. Email: {dto.Email}");
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:20]: better write message like $,,User with this email {your email} does not exist,,

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        var token = await _userManager.GeneratePasswordResetTokenAsync(user);
+
+        return token;
+    }
+
+    public async Task<bool> IsResetPassword(ResetPasswordDTO dto)
+    {
+        var user = await _userManager.FindByEmailAsync(dto.Email);
+
+        if (user == null)
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:31]: same with ??

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+            throw new NotFoundException($"Unable to find user by specified email. Email: {dto.Email}");
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:32]: also, some other message

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+        var isSamePassword = await _userManager.CheckPasswordAsync(user, dto.NewPassword);
+
+        if (isSamePassword)
+            throw new IncorrectParametersException("New password have to differ from the old one");
+
+        var result = await _userManager.ResetPasswordAsync(user, dto.ResetToken, dto.NewPassword);
+
+        return result.Succeeded;
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:41]: where is validation if result.Succeeded == false?

File: StudyHub/StudyHub.BLL/Services/Auth/TokenService.cs [hunk 14]
```diff
@@ -1,58 +1,30 @@
...
 using StudyHub.Common.Models;
 using StudyHub.Entities;
 using System.IdentityModel.Tokens.Jwt;
 using System.Security.Claims;
 using System.Security.Cryptography;
 using System.Text;
 
-namespace StudyHub.BLL.Services;
-
+namespace StudyHub.BLL.Services.Auth;
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/TokenService.cs:12]: empty line

File: StudyHub/StudyHub.Common/DTO/AuthDTO/ResetPasswordDTO.cs [hunk 15]
```diff
@@ -0,0 +1,8 @@
+﻿namespace StudyHub.Common.DTO.AuthDTO;
```
Comment (Podgore) [inline StudyHub/StudyHub.Common/DTO/AuthDTO/ResetPasswordDTO.cs:1]: empty line

File: StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs:1]: Honestly, I dont think that we needed validation for refresh token here. Meaby take it on separate file?

File: StudyHub/StudyHub.BLL/Services/Interfaces/Auth/IAuthService.cs [hunk 16]
```diff
@@ -0,0 +1,9 @@
+﻿using StudyHub.Common.DTO.AuthDTO;
+
+namespace StudyHub.BLL.Services.Interfaces.Auth;
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Interfaces/Auth/IAuthService.cs:3]: empty line

File: StudyHub/StudyHub.BLL/Services/Interfaces/Auth/IGoogleAuthService.cs [hunk 17]
```diff
@@ -0,0 +1,8 @@
+﻿using StudyHub.Common.DTO.AuthDTO;
+
+namespace StudyHub.BLL.Services.Interfaces.Auth;
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Interfaces/Auth/IGoogleAuthService.cs:3]: empty line

File: StudyHub/StudyHub.BLL/Services/Interfaces/Auth/IPasswordService.cs [hunk 18]
```diff
@@ -0,0 +1,8 @@
+﻿using StudyHub.Common.DTO.AuthDTO;
+
+namespace StudyHub.BLL.Services.Interfaces.Auth;
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Interfaces/Auth/IPasswordService.cs:3]: empty line

File: StudyHub/StudyHub.BLL/Services/Interfaces/Auth/ITokenService.cs [hunk 19]
```diff
@@ -0,0 +1,8 @@
+﻿using StudyHub.Entities;
+
+namespace StudyHub.BLL.Services.Interfaces.Auth;
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Interfaces/Auth/ITokenService.cs:3]: empty line

File: StudyHub/StudyHub.Seeding/Behaviours/UsersSeedingBehaviour.cs
Comment (Podgore) [inline StudyHub/StudyHub.Seeding/Behaviours/UsersSeedingBehaviour.cs:1]: look, I dont think that we extremly need seeding, thing about it

**Review by Podgore — CHANGES_REQUESTED** [review]

File: StudyHub/StudyHub.Seeding/Behaviours/UsersSeedingBehaviour.cs
Comment (Podgore) [reply StudyHub/StudyHub.Seeding/Behaviours/UsersSeedingBehaviour.cs:1]: doesn't matter, your right

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs (hunk 7 above)
Comment (fedyafedko) [reply StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:19]: why?

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs (hunk 10 above)
Comment (fedyafedko) [reply StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:20]: In all our project, we used to write error messages exactly like that, so I think it's better to leave it like that

Comment (aquaprogit) [reply StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:20]: I think @fedyafedko's variant quite clarify, that's why don't need changes here.

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs (hunk 7 above)
Comment (aquaprogit) [reply StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:19]: > why?

I think this way ForgotPassword (and add suffix **Async** to it btw) will look much simpler or at least shorter, idk

File: StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs
Comment (aquaprogit) [reply StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs:1]: I think extension method fits here just fine

File: StudyHub/StudyHub.BLL/Profiles/UserProfile.cs [hunk 20]
```diff
@@ -7,11 +8,15 @@ namespace StudyHub.BLL.Profiles;
...
     {
         CreateMap<InvitedUserDTO, InvitedUser>();
         CreateMap<RegisterUserDTO, User>()
-            .ForMember(dest => dest.UserName, opt => opt.MapFrom(src => src.Email))
-            .ForMember(dest => dest.EmailConfirmed, opt => opt.MapFrom(_ => true));
+            .ForMember(dest => dest.EmailConfirmed, opt => opt.MapFrom(_ => true))
+            .ForMember(dest => dest.UserName, opt => opt.MapFrom(src => src.Email));
+
+        CreateMap<GoogleJsonWebSignature.Payload, User>()
+            .ForMember(dest => dest.Email, opt => opt.MapFrom(src => src.Email))
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Profiles/UserProfile.cs:19]: Also nedeed delete map here

File: StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs [hunk 21]
```diff
@@ -1,77 +1,67 @@
...
         _tokenService = tokenService;
+        _mapper = mapper;
     }
 
-    public async Task<AuthSuccessDTO> LoginAsync(LoginUserDTO user)
+    public async Task<AuthSuccessDTO> LoginAsync(LoginUserDTO dto)
     {
-        var existingUser = await _userManager.FindByEmailAsync(user.Email)
-            ?? throw new NotFoundException($"Unable to find user by specified email. Email: {user.Email}");
+        var user = await _userManager.FindByEmailAsync(dto.Email);
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs:32]: also add here ,,?? throw new .... ,,

File: StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs [hunk 22]
```diff
@@ -1,77 +1,67 @@
...
     public async Task<AuthSuccessDTO> RegisterAsync(RegisterUserDTO dto)
     {
-        var invitedUser = _invitedUserRepository.FirstOrDefault(e => e.Email == dto.Email)
-            ?? throw new NotFoundException($"User with this email wasn't invited: {dto.Email}");
+        var user = await _userManager.FindByEmailAsync(dto.Email);
 
-        if (invitedUser.Token != dto.Token)
-            throw new IncorrectParametersException("Passed token isn't valid.");
+        if (user != null)
+            throw new NotFoundException($"User with specified email already exists. Email: {dto.Email}");
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs:50]: Why if we fing user throws exeptions NotFound?

File: StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs [hunk 23]
```diff
@@ -1,77 +1,67 @@
...
 
-        var user = _mapper.Map<User>(dto);
+        user = _mapper.Map<User>(dto);
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs:52]: it seems weird for me that we used one ,,user,, for validation and, subsequently, registaring them

File: StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs [hunk 24]
```diff
@@ -91,16 +81,15 @@ public async Task<AuthSuccessDTO> RefreshTokenAsync(string accessToken, string r
...
-        return await GetAuthTokens(user);
+        return await GetAuthTokensAsync(user);
     }
 
-    private async Task<AuthSuccessDTO> GetAuthTokens(User user)
+    protected async Task<AuthSuccessDTO> GetAuthTokensAsync(User user)
     {
-        var roles = (await _userManager.GetRolesAsync(user)).ToArray();
-        return new AuthSuccessDTO(_tokenService.GenerateJwtToken(user!, roles),
+        return new AuthSuccessDTO(await _tokenService.GenerateJwtTokenAsync(user!),
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs:92]: I dont get why delete roles from there?

File: StudyHub/StudyHub.BLL/Extensions/TokensExtensions.cs
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Extensions/TokensExtensions.cs:1]: Seek advice with @aquaprogit, It will be nice if fuction GetPrincipalFromToken be here

**Review by Podgore — CHANGES_REQUESTED** [review]

File: StudyHub/StudyHub.BLL/Extensions/TokensExtensions.cs
Comment (aquaprogit) [reply StudyHub/StudyHub.BLL/Extensions/TokensExtensions.cs:1]: It's bad practice to have overloading extensions on one concrete usage of it to that global type "string".

File: StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs (hunk 24 above)
Comment (aquaprogit) [reply StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs:92]: We don't need roles declaration to be found that high away from usage

File: StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs (hunk 23 above)
Comment (fedyafedko) [reply StudyHub/StudyHub.BLL/Services/Auth/AuthService.cs:52]: Why should we create another one?

**Review by Podgore — APPROVED** [review]

File: StudyHub/StudyHub.Common/DTO/AuthDTO/ForgotPasswordDTO.cs [hunk 25]
```diff
@@ -0,0 +1,6 @@
+﻿namespace StudyHub.Common.DTO.AuthDTO;
+
+public class ForgotPasswordDTO
+{
+    public string Email{ get; set; } = string.Empty;
```
Comment (aquaprogit) [inline StudyHub/StudyHub.Common/DTO/AuthDTO/ForgotPasswordDTO.cs:5]: formatting

File: StudyHub/StudyHub.Common/DTO/AuthDTO/ResetPasswordDTO.cs [hunk 26]
```diff
@@ -0,0 +1,9 @@
...
+
+public class ResetPasswordDTO
```
Comment (aquaprogit) [inline StudyHub/StudyHub.Common/DTO/AuthDTO/ResetPasswordDTO.cs:3]: ResetPasswordRequest

File: StudyHub/StudyHub.Common/DTO/AuthDTO/ForgotPasswordDTO.cs [hunk 27]
```diff
@@ -0,0 +1,6 @@
+﻿namespace StudyHub.Common.DTO.AuthDTO;
+
+public class ForgotPasswordDTO
```
Comment (aquaprogit) [inline StudyHub/StudyHub.Common/DTO/AuthDTO/ForgotPasswordDTO.cs:3]: ForgotPasswordRequest

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        var user = await _userManager.FindByEmailAsync(dto.Email)
+            ?? throw new NotFoundException($"User with this email does not exist. Email: {dto.Email}");
+
+        var token = await _userManager.GeneratePasswordResetTokenAsync(user);
+
+        return token;
+    }
+
+    public async Task<bool> IsResetPassword(ResetPasswordDTO dto)
```
Comment (aquaprogit) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:28]: ResetPasswordAsync

File: StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+public class PasswordService : IPasswordService
+{
+    private readonly UserManager<User> _userManager;
+
+    public PasswordService(UserManager<User> userManager)
+    {
+        _userManager = userManager;
+    }
+
+    public async Task<string> ForgotPassword(ForgotPasswordDTO dto)
```
Comment (aquaprogit) [inline StudyHub/StudyHub.BLL/Services/Auth/PasswordService.cs:18]: ForgotPasswordAsync

File: StudyHub/StudyHub/Controllers/AuthController.cs [hunk 30]
```diff
@@ -9,30 +9,48 @@ namespace StudyHub.Controllers;
...
-    public AuthController(IAuthService authService)
+    public AuthController(IAuthService authService, IPasswordService passwordService)
     {
         _authService = authService;
+        _passwordService = passwordService;
     }
 
-    [HttpPost("[action]")]
-    public async Task<IActionResult> Register(RegisterUserDTO userDTO)
+    [HttpPost("sign-up")]
```
Comment (aquaprogit) [inline StudyHub/StudyHub/Controllers/AuthController.cs:20]: #47 should remove this changes, so you can implement it if @Podgore don't want easy _points_

File: StudyHub/StudyHub/Controllers/GoogleAuthController.cs [hunk 31]
```diff
@@ -14,10 +14,17 @@ public GoogleAuthController(IGoogleAuthService googleAuthService)
...
+        var result = await _googleAuthService.GoogleRegisterAsync(authorizationCode);
+        return Ok(result);
+    }
+
     [HttpPost("sign-in")]
     public async Task<IActionResult> GoogleSignIn([FromHeader(Name = "Authorization-Code")] string authorizationCode)
     {
-        var result = await _googleAuthService.GoogleLogin(authorizationCode);
+        var result = await _googleAuthService.GoogleLoginAsync(authorizationCode);
         return Ok(result);
```
Comment (aquaprogit) [inline StudyHub/StudyHub/Controllers/GoogleAuthController.cs:28]: consistent naming for service and controller methods. Better to have Register and Login, I guess

File: StudyHub/StudyHub/Controllers/SubjectController.cs [hunk 32]
```diff
@@ -54,11 +54,11 @@ public async Task<IActionResult> GetSubject(Guid subjectId)
     }
 
     [HttpGet("teacher")]
-    [Authorize]
-    public IActionResult GetSubjectForTeacher()
+    [Authorize(Roles ="Teacher")]
```
Comment (aquaprogit) [inline StudyHub/StudyHub/Controllers/SubjectController.cs:57]:
```suggestion
    [Authorize(Roles = "Teacher")]
```

File: StudyHub/StudyHub/StudyHub.csproj [hunk 33]
```diff
@@ -30,6 +30,7 @@
 		</PackageReference>
 		<PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="8.0.0" />
 		<PackageReference Include="NuGet.Protocol" Version="6.7.0" />
+		<PackageReference Include="Scrutor" Version="4.2.2" />
```
Comment (aquaprogit) [inline StudyHub/StudyHub/StudyHub.csproj:33]: ?

**Review by aquaprogit — CHANGES_REQUESTED** [review]

**Review by Podgore — APPROVED** [review]

**Review by aquaprogit — APPROVED** [review]

**Pull request merged**